infrastructure/comms/src/connection/net_address/mod.rs

```diff
@@ -41,51 +59,90 @@ pub enum NetAddress {
     I2P(I2PAddress),
 }
 
+impl NetAddress {
+    /// Returns true if the [`NetAddress`] is an IP address, otherwise false
+    pub fn is_ip(&self) -> bool {
+        match *self {
+            NetAddress::IP(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if the [`NetAddress`] is a Tor Onion address, otherwise false
+    pub fn is_tor(&self) -> bool {
+        match *self {
+            NetAddress::Tor(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if the [`NetAddress`] is an I2P address, otherwise false
+    pub fn is_i2p(&self) -> bool {
+        match *self {
+            NetAddress::I2P(_) => true,
+            _ => false,
+        }
+    }
+}
+
 impl FromStr for NetAddress {
     type Err = NetAddressError;
 
+    /// Parses a [`str`] into a [`NetAddress`]
     fn from_str(address: &str) -> Result<Self, Self::Err> {
         if let Ok(addr) = address.parse::<SocketAddress>() {
-            Ok(NetAddress::IP(addr))
+            Ok(addr.into())
         } else if let Ok(addr) = address.parse::<OnionAddress>() {
-            Ok(NetAddress::Tor(addr))
+            Ok(addr.into())
         } else if let Ok(addr) = address.parse::<I2PAddress>() {
-            Ok(NetAddress::I2P(addr))
+            Ok(addr.into())
         } else {
             Err(NetAddressError::ParseFailed)
         }
     }
 }
 
+impl From<SocketAddress> for NetAddress {
```

Comment:
This might  be unsafe, what happens when the given given address is incorrect?

Reply:
What do you mean by 'incorrect'? You should not be able to create a malformed `SocketAddress`